### PR TITLE
feat(dev-env): Add a concept of init-only containers

### PR DESCRIPTION
**Pull request by sjinks**
## Description

This PR adds a concept of "init-only" containers: they run some initialization code and the exit. This allows us to save some RAM and CPU cycles.

## Steps to Test

a. The CI should pass.
b. Apply the patch, create a new dev environment from scratch, play with it, stop it, restart it, destroy it - everything should work.

### CODE REVIEW

File: assets/dev-env.lando.template.yml.ejs [hunk 1]
```diff
@@ -154,6 +156,10 @@ services:
           target: /scripts
           volume:
             nocopy: true
+#      environment:
+#        LANDO_NO_SCRIPTS: 1
+#        LANDO_NEEDS_EXEC: 1
+#    initOnly: true
```
Comment (sjinks) [inline assets/dev-env.lando.template.yml.ejs:170]: The commented-out lines will need to be uncommented in 2.29.

Comment (pschoffer) [reply assets/dev-env.lando.template.yml.ejs:170]: Is this still the case. Could you add more context to this please?

File: src/lib/dev-environment/dev-environment-lando.js [hunk 2]
```diff
@@ -98,6 +98,25 @@ async function getLandoConfig() {
 
 const appMap: Map< string, App > = new Map();
 
+async function getApplication( lando: Lando, instancePath: string ): Promise< App > {
```
Comment (pschoffer) [inline src/lib/dev-environment/dev-environment-lando.js]: Could we somehow make the difference between getApplication and getLandoApplication more clear? What does the `getLandoApplication` actually do?

**Review by pschoffer — CHANGES_REQUESTED** [review]
This is awesome, sorry it took me that long to get to this @sjinks

**Comment by sjinks** [comment]
Let's postpone this one a bit because I want to finish the conversion of dev-env to TypeScript first, for I am sick and tired of resolving conflicts here and there :-)

File: assets/dev-env.lando.template.yml.ejs (hunk 1 above)
Comment (sjinks) [reply assets/dev-env.lando.template.yml.ejs:170]: `vip dev-env start` re-pulls images from the repo once a week. I needed to give some time for the required changes to propagate to the existing environments.

File: src/lib/dev-environment/dev-environment-lando.js (hunk 2 above)
Comment (sjinks) [reply src/lib/dev-environment/dev-environment-lando.js]: I have renamed one of them.

Initially, there used to be only `getLandoApplication()`. It ensured that we initialize every application only once. Then we added a feature to recover broken landofiles. The common part was extracted into `getApplication()` (which is now `initLandoApplication()`).

File: assets/dev-env.lando.template.yml.ejs (hunk 1 above)
Comment (sjinks) [reply assets/dev-env.lando.template.yml.ejs:170]: I failed to sort out the issue with permissions, will leave this for another PR.

**Comment by pschoffer** [comment]
I get following when starting existing environment:

```
Scanning to determine which services are ready... Please stand by...
Waiting for core files to be copied
Already existing wp-config.php file
Waiting for MySQL to come online
Checking for database connectivity...
Copying dev-env-plugin.php to mu-plugins
Waiting for Elasticsearch to come online........
Checking for WordPress installation...
Success: The cache was flushed.
WordPress already installed
Error: Cannot read properties of null (reading 'length')

Please re-run the command with "--debug @automattic/vip:bin:dev-environment" appended to it and provide the stack trace on the support ticket.

Example:

vip dev-env <command> <arguments> --debug @automattic/vip:bin:dev-environment 

```

I'll try to debug some more later today.

EDIT: This was unrelated bug related to elasticsearch but I added oneliner to fix it here. I hope thats ok @sjinks

**Review by pschoffer — APPROVED** [review]

**Comment by sjinks** [comment]
I have applied some TypeScript magic to get rid of ESLint suppression.

**Pull request merged**